p256/src/arithmetic/field/field32.rs

```diff
@@ -42,11 +42,36 @@ pub(super) const fn add(a: U256, b: U256) -> U256 {
 /// Multiplies by a single-limb integer.
 /// Multiplies the magnitude by the same value.
 pub(super) fn mul_single(a: U256, rhs: u32) -> U256 {
-    let mut result = U256::ZERO;
-    for _i in 0..rhs {
-        result = add(result, a)
-    }
-    result
+    let a_limbs = a.as_limbs();
+    let rhs_limb = Limb::from_u32(rhs);
+    let (w0, carry) = a[0].mac(b[0], Limb::ZERO);
+    let (w1, carry) = a[1].mac(b[1], carry);
+    let (w2, carry) = a[2].mac(b[2], carry);
+    let (w3, carry) = a[3].mac(b[3], carry);
+    let (w4, carry) = a[4].mac(b[4], carry);
+    let (w5, carry) = a[5].mac(b[5], carry);
+    let (w6, carry) = a[6].mac(b[6], carry);
+    let (w7, w8) = a[7].mac(b[7], carry);
+    // Attempt to subtract the modulus, to ensure the result is in the field.
+    let modulus = MODULUS.0.as_limbs();
+
+    let (result, _) = sub_inner(
```

Comment:
I don't think this step works correctly for a `multiply`, even though it worked for `add`. This `sub_inner` call subtracts off the modulus, and then the internal logic of sub_inner uses the highest-order limb as a `bitand` mask with the modulus and adds that back on. In the case of `add`, this works well -- the highest order limb can be either 0 or 1 after the addition, and once the modulus is subtracted off it can only be 0 or -1 (in the normalized inputs case, anyway) -- and -1 is treated as 0xffffffff. But in this case, the highest order limb can be almost anything (for instance, consider what happens if you multiply `modulus - 1` by `2^30`). Then the `sub_inner` logic will be `bitand`ing the modulus with some arbitrary u32, and adding that back on. This will produce a result not at all related to the intended multiply.

Reply:
I see what you mean here. Thanks for pointing out the difference! I have attempted to change it into the implementation that you have suggested previously with the help of [`U256::const_rem`](https://docs.rs/crypto-bigint/latest/crypto_bigint/struct.Uint.html#method.const_rem) function
1. First track the carries
2. Reduce the carries by modding with prime
3. Modular addition of reduced carries with non-carries
```rust 
pub(super) fn mul_single(a: U256, rhs: u32) -> U256 {
    let a_limbs = a.as_limbs();
    let rhs_limb = Limb::from_u32(rhs);
    let (w0, carry) = Limb::ZERO.mac(a_limbs[0], rhs_limb, Limb::ZERO);
    let (w1, carry) = Limb::ZERO.mac(a_limbs[0], rhs_limb, carry);
    let (w2, carry) = Limb::ZERO.mac(a_limbs[0], rhs_limb, carry);
    let (w3, w4) = Limb::ZERO.mac(a_limbs[0], rhs_limb, carry);

    // Reduce the carry mod prime
    let carry = U256::from(w4);
    let (reduced_carry, _) = carry.const_rem(&MODULUS.0);

    // Modular addition of non-carry and reduced carry
    let non_carries = U256::new([w0, w1, w2, w3]);
    add(non_carries, reduced_carry)
}
```

Reply:
Hmm, I think this doesn't quite work either -- `w4` is a 64-bit limb here (or 32-bit limb in the other case), and so setting `carry = U256::from(w4)` gets you something already smaller than `MODULUS` (and so the reduce is a no-op) but not in the right way -- it's reduced by 2^256, not the modulus.

I have two approaches in mind that I think could work (there are probably more):

1. Create `carry` as a larger type (`U512`?), set it to `w4 << 256`, and reduce
2. Note that since 2^256 is slightly larger than `MODULUS`, you know that 2^256 reduced by `MODULUS` is `2^256 - MODULUS`. Thus, `w4 << 256` mod `MODULUS` is just `w4 * (2^256 - MODULUS)`. This won't even need to be reduced, as both `w4` and `2^256 - MODULUS` fit within a limb, and thus their product fits within 2 limbs i.e. is less than the modulus.

Reply:
I am looking more at the second approach because I am not too sure how we can define a larger type for the carry if we were to continue using the `mac` function.

However, I have some clarifications regarding the second approach. The `MODULUS` of p256 `= 2^{224}(2^{32} − 1) + 2^{192} + 2^{96} − 1` and from my calculations I believe `2^256 - MODULUS ` is equal to `0x00000000FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFF000000000000000000000001` in hexadecimal. Which, when converted to limbs will not fit within a single limb.

From your second approach, I was thinking of implementing something like this in code. After tracking the carries using `mac`, we can first get `2^256 - MODULUS` which is the hex value and convert it into limbs
```rust
// Define 2^256 - MODULUS
let subtracted_result_str: &str =
    "00000000FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFF000000000000000000000001";

let subtracted_result = U256::from_be_hex(subtracted_result_str);
let subtracted_result_limb = subtracted_result.as_limbs();
```
This could potentially be done in some other way like using `Limb::u32` or basically just not using the hex value, however, I am not too sure how to achieve that yet so if you have any suggestions do let me know. 
After getting `2^256 - MODULUS`, and assuming this fits into a single limb, we can probably just get the product of `w4 * (2^256 - MODULUS)`
```rust
let (carry1, carry2) = Limb::ZERO.mac(w4, subtracted_result_limb.[0], Limb::ZERO); 
```
which should fit within 2 limbs `carry1` and `carry2`. And to do a modular addition with the non-carries we can
```rust
let carries = U256::new([Limb::ZERO, Limb::ZERO, carry2, carry1]);
let non_carries = U256::new([w0, w1, w2, w3]);
add(carries, non_carries)
```
So the issue here I guess is still how does the value `(2^256 - MODULUS)` fit within a single limb. Also, even if it fits within a single limb for u64, will it fit within a single limb for u32? Perhaps, my calculations were wrong and I really hope you could shed more light on this and provide some examples of how you would go about coding this? Thanks for you help so far! 😄

Reply:
> the issue here I guess is still how does the value (2^256 - MODULUS) fit within a single limb.

Oh, good call, this was my mistake, I thought `2^256 - MODULUS` was smaller than it actually is. It looks like actually `2^256 - MODULUS` needs 7 limbs (of 32-bits) to fit. For the 32-bit case, this will work out pretty well; for the 64-bit case, it's going to be more awkward.

For the 32-bit case, we can compute the carry component using the existing multiplication algorithm but ignoring the final carry, because 1 limb times 7 limbs gives us at most 8 limbs. It might be necessary to split off a `mul_inner` that we use for this purpose, but fundamentally, we can reuse the existing limbed multiply without needing to worry about the uppermost carry, then do a modular 256-bit addition to combine this with the non-carry result.

For the 64-bit case it's messier. I suppose you can do the same approach as the 32-bit case, but with two rounds of the `mul_inner` -- the very first time through you have an overflow carry of `w4` that might take the full 64 bits; then do the approach of the 32-bit case, but now keep track of the `w4` carry because it might have up to 32 bits in it; now do the approach of the 32-bit case again, but because the incoming carry can't be more than 32 bits and the `2^256 - MODULUS` fits in 224 bits, their product will fit in 256 bits and the `w4` could be dropped. This will need some careful thought for how to combine all three parts (uncarried, partially carried, fully carried), but I _think_ it should all be possible to work out. Definitely double-check my logic here though, this 64-bit case is messy.

Reply:
So for the implementation, 32 bit case it is pretty straight forward since the defined `mul_inner` will just return a reduced carry that fits in at most 8 limbs using the same multiplication logic we have
```rust
fn mul_inner(a: U256, b: Limb) -> U256 {
    let a_limbs = a.as_limbs();
    let (w0, carry) = Limb::ZERO.mac(a_limbs[0], b, Limb::ZERO);
    let (w1, carry) = Limb::ZERO.mac(a_limbs[1], b, carry);
    let (w2, carry) = Limb::ZERO.mac(a_limbs[2], b, carry);
    let (w3, cary) = Limb::ZERO.mac(a_limbs[3], b, carry);
    let (w4, carry) = Limb::ZERO.mac(a_limbs[4], b, carry);
    let (w5, carry) = Limb::ZERO.mac(a_limbs[5], b, carry);
    let (w6, carry) = Limb::ZERO.mac(a_limbs[6], b, carry);
    // We can ignore the last carry
    let (w7, _) = Limb::ZERO.mac(a_limbs[7], b, carry);

    U256::new([w0, w1, w2, w3, w4, w5, w6, w7])
}
```
and we call `mul_inner` to get a reduced carry from our initial `mul_single` function
```rust
// Define 2^256 - MODULUS
let subtracted_result_str: &str =
    "00000000FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFF000000000000000000000001";

let subtracted_result = U256::from_be_hex(subtracted_result_str);
// Calculate w8 << 2^256 = w8 * (w^256 - MODULUS)
let reduced_carry = mul_inner(subtracted_result, w8);

// Modular addition of non-carry and reduced carry
let non_carries = U256::new([w0, w1, w2, w3, w4, w5, w6, w7]);
add(non_carries, reduced_carry)
```
For the 64 bit case `mul_inner` is defined differently. Once we get the initial carry of `w4` (which could take up to 64 bits) in `mul_single`, we calculate `w4 * (2^256 - MODULUS)` but this operation could result in an overflow of up to 32 bits hence in `mul_inner` we need to track this overflow which is again defined as `w4`. Using this overflow of at most 32 bits, we calculate `w4 * (2^256 - MODULUS)` again but this time the result should fit in 4 limbs `[c0..c3]` since `32 + 224 = 256 bits` at most, hence ignoring the last carry. Finally we perform modular addition the "inner" carries and non_carries to get the `reduced_carry` needed for the initial `mul_single`. Therefore this was how I defined 64 bit `mul_inner`
```rust
fn mul_inner(a: U256, b: Limb) -> U256 {
    let a_limbs = a.as_limbs();
    let (w0, carry) = Limb::ZERO.mac(a_limbs[0], b, Limb::ZERO);
    let (w1, carry) = Limb::ZERO.mac(a_limbs[1], b, carry);
    let (w2, carry) = Limb::ZERO.mac(a_limbs[2], b, carry);
    let (w3, w4) = Limb::ZERO.mac(a_limbs[3], b, carry);
    let non_carries = U256::new([w0, w1, w2, w3]);

    let (c0, carry) = Limb::ZERO.mac(a_limbs[0], w4, Limb::ZERO);
    let (c1, carry) = Limb::ZERO.mac(a_limbs[1], w4, carry);
    let (c2, carry) = Limb::ZERO.mac(a_limbs[2], w4, carry);
    let (c3, _) = Limb::ZERO.mac(a_limbs[3], w4, carry);
    let reduced_carry = U256::new([c0, c1, c2, c3]);

    add(non_carries, reduced_carry)
}
```
`mul_single` in 64 bits is defined similarly to that of 32 bits. Mathematically, the logic and the bits look sound to me! Although I am not too certain the part of combining uncarried, partially carried, fully carried that you mentioned, I think what it corresponds to here is:
- uncarried = `non_carries` in initial `mul_single`
- partially carried = `non_carries` in `mul_inner`
- fully carried = `reduced_carries` in `mul_inner`

Reply:
Sorry for the delayed response, I had an inopportunely timed case of covid. Doing better now though!

The 32-bit case makes sense to me. I don't think the 64-bit case is right though -- it looks like you are multiplying the carry portion by `a` a second time, and I think you need to be multiplying it by `0x00000000FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFF000000000000000000000001`.

For the 3 parts to add back together, I think they are `non_carries`, `reduced_carry`, and `_` from `mul_inner`, except `_` is a carry limb and needs to be multiplied by `0x00000000FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFF000000000000000000000001` again to reduce it into 256 bits.

Reply:
Thanks for your patience!

I think I understand now and that your 64-bit mul_single is correct. I think that the `mul_inner` only works because `a` is specifically 2^256 - P, so what do you think about defining the constant `00000000FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFF000000000000000000000001` in the `mul_inner` algorithm instead of passing it in as `a` -- that `mul_inner` function won't do anything sensible with a different `a`, right?

Also, while I think I have followed the reasoning and it works correctly, it would be nice to have some tests that check the calculations too so we don't have to rely on my reasoning. Perhaps validating `mul_single` on 2^224 * 2^31 (for no carries), 2^224*2^32 (carrying 1), 2^224 * 2^33 (a carry that fits in 32 bits), and 2^250 * 2^50 (a carry that doesn't fit in 32 bits) or similar would be good test inputs, as they test the major paths of the `mul_single`.